nix/nix-build.sh

```diff
@@ -32,14 +32,30 @@ then
     nix-shell --pure $SHELL_NIX_FILE
 else
     ARGS=(--arg crates "${CRATES}" --arg singleStep "${SINGLE_STEP}" --arg rustflags "${RUSTFLAGS}")
-    nix-build --max-jobs auto $SHELL_NIX_FILE "${ARGS[@]}"
+
+    # first we download all dependencies
+    echo fetching depedencies...
+    CARGO_HOME="$(realpath ~/cargo-home)"
+    nix-shell --pure --run "CARGO_HOME=$CARGO_HOME cargo fetch --locked --offline || CARGO_HOME=$CARGO_HOME cargo fetch --locked"
+
+    echo building...
+    nix-build --max-jobs auto --option sandbox true --arg cargoHomePath "$CARGO_HOME" --show-trace $SHELL_NIX_FILE "${ARGS[@]}"
+    echo build finished
+
+    echo copying results...
```

Comment:
I do not quite understand what we're trying to achieve in this copy section, could you please elaborate?